modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItIndexAndIndexStorageDestructionTest.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.ignite.internal.table.distributed.TableSchemaAwareIndexStorage;
+import org.apache.ignite.lang.Cursor;
+import org.apache.ignite.sql.SqlException;
+import org.apache.ignite.table.KeyValueView;
+import org.apache.ignite.table.Tuple;
+import org.apache.ignite.table.criteria.CriteriaException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests about accessing destroyed index storages.
+ */
+class ItIndexAndIndexStorageDestructionTest extends ClusterPerTestIntegrationTest {
+    private static final String TABLE_NAME = "TEST_TABLE";
+    private static final String INDEX_NAME = "TEST_INDEX";
+
+    private static final int PREEXISTING_KEY = 1;
+    private static final int ANOTHER_KEY = 2;
+
+    private IgniteImpl node;
+
+    @Override
+    protected int initialNodes() {
+        return 1;
+    }
+
+    @BeforeEach
+    void createTableAndIndex() {
+        node = cluster.aliveNode();
+
+        cluster.doInSession(0, session -> {
+            session.execute(null, "CREATE TABLE " + TABLE_NAME + " (id INT PRIMARY KEY, name VARCHAR)");
+            session.execute(null, "CREATE INDEX " + INDEX_NAME + " ON " + TABLE_NAME + "(name)");
+
+            session.execute(null, "INSERT INTO " + TABLE_NAME + " (id, name) VALUES (" + PREEXISTING_KEY + ", 'John')");
+        });
+
+        TableImpl table = (TableImpl) node.tables().table(TABLE_NAME);
+
+        initiateIndexStoragesDestruction(table, INDEX_NAME);
+    }
+
+    private void initiateIndexStoragesDestruction(TableImpl table, String indexName) {
+        int indexId = indexId(indexName);
+
+        for (int partitionNumber = 0; partitionNumber < table.internalTable().partitions(); partitionNumber++) {
+            TableIndexStoragesSupplier storagesSupplier = table.indexStorageAdapters(partitionNumber);
+            TableSchemaAwareIndexStorage schemaAwareStorage = storagesSupplier.get().get(indexId);
+            IndexStorage indexStorage = schemaAwareStorage.storage();
+
+            if (indexStorage instanceof AbstractPageMemoryIndexStorage) {
+                ((AbstractPageMemoryIndexStorage<?, ?>) indexStorage).transitionToDestroyedState();
+            } else if (indexStorage instanceof AbstractRocksDbIndexStorage) {
+                ((AbstractRocksDbIndexStorage) indexStorage).transitionToDestroyedState();
+            } else {
+                fail("Index storage is not of a type supported by the test: " + indexStorage.getClass());
+            }
+        }
+    }
+
+    private int indexId(String indexName) {
+        CatalogIndexDescriptor indexDescriptor = node.catalogManager().aliveIndex(indexName, node.clock().nowLong());
+        assertThat(indexDescriptor, is(notNullValue()));
+
+        return indexDescriptor.id();
+    }
+
+    @Test
+    void insertTouchingDestroyedIndexStorageDoesNotFailWholeWrite() {
+        KeyValueView<Integer, String> view = node.tables().table(TABLE_NAME)
```

Comment:
How do you check that this call touches a destroyed index?

Reply:
Line 105 doesn't touch it. Code inside `assertDoesNotThrow()` does. I wrote the test before doing the fix, the test failed due to an exception about a destroyed storage.

Reply:
> Code inside assertDoesNotThrow() does. I wrote the test before doing the fix, the test failed due to an exception about a destroyed storage.

I understand, but this looks like a fragile approach. I would expect that we would validate somehow, that the destroyed index is indeed accessed,

Reply:
The index is destroyed (I destroyed the storages explicitly). There are tests that make sure that a destroyed storage throws exceptions. Then I test the system via API. No exception flyies (and it flied before the fix). I don't see what else should be done.

Reply:
I added a unit test for `SortedIndexLocker`